compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/FlattenTuplesInCFG.cpp

```diff
@@ -101,126 +101,173 @@ Value recursiveRetuple(Type oldType, Operation::result_range *values,
   }
+};
 
-  return success();
-}
+class DetupleCallOp : public OpRewritePattern<func::CallOp> {
+  using OpRewritePattern::OpRewritePattern;
 
-LogicalResult convertOp(func::CallIndirectOp oldOp, OpBuilder &builder,
-                        IRMapping &mapping) {
-  llvm::SmallVector<Value> newArgs;
-  if (failed(untupleAndLookupValues(oldOp.getOperands(), newArgs, builder,
-                                    oldOp.getLoc(), mapping))) {
-    return failure();
-  }
+  LogicalResult matchAndRewrite(func::CallOp oldOp,
+                                PatternRewriter &builder) const override {
+    bool isTuple = false;
+    for (auto operand : oldOp.getOperands())
+      isTuple |= isa<TupleType>(operand.getType());
 
-  auto newOp = builder.create<func::CallIndirectOp>(oldOp.getLoc(),
-                                                    oldOp.getCallee(), newArgs);
-  copyOperationAttrs(oldOp, newOp);
+    for (auto result : oldOp.getResults())
+      isTuple |= isa<TupleType>(result.getType());
 
-  for (auto [oldResult, newResult] :
-       llvm::zip_equal(oldOp.getResults(), newOp.getResults())) {
-    mapping.map(oldResult, newResult);
-  }
+    if (!isTuple)
+      return builder.notifyMatchFailure(oldOp, "No detupling required");
 
-  return success();
-}
+    llvm::SmallVector<Value> newArgs;
+    if (failed(untupleAndLookupValues(oldOp.getOperands(), newArgs, builder,
+                                      oldOp.getLoc()))) {
+      return builder.notifyMatchFailure(oldOp, "failed to untuple values");
+    }
+
+    SmallVector<Type> resultTypes;
+    untupleTypes(oldOp.getResultTypes(), resultTypes);
+    auto newOp = builder.create<func::CallOp>(
+        oldOp->getLoc(), oldOp.getCallee(), resultTypes, newArgs);
+    copyOperationAttrs(oldOp, newOp);
+
+    auto newResults = newOp.getResults();
+    llvm::SmallVector<Value> retupledResults;
+    for (auto oldResult : oldOp.getResults()) {
+      auto newResult = recursiveRetuple(oldResult.getType(), &newResults,
+                                        builder, oldOp->getLoc());
+      retupledResults.push_back(newResult);
+    }
 
-LogicalResult convertOp(cf::BranchOp oldOp, OpBuilder &builder,
-                        IRMapping &mapping) {
-  llvm::SmallVector<Value> newArgs;
-  if (failed(untupleAndLookupValues(oldOp.getOperands(), newArgs, builder,
-                                    oldOp.getLoc(), mapping))) {
-    return failure();
+    builder.replaceOp(oldOp, retupledResults);
+    return success();
   }
+};
 
-  auto newOp = builder.create<cf::BranchOp>(
-      oldOp.getLoc(), mapping.lookupOrNull(oldOp.getDest()), newArgs);
+class DetupleIndirectCallOp : public OpRewritePattern<func::CallIndirectOp> {
+  using OpRewritePattern::OpRewritePattern;
 
-  copyOperationAttrs(oldOp, newOp);
-  return success();
-}
+  LogicalResult matchAndRewrite(func::CallIndirectOp oldOp,
+                                PatternRewriter &builder) const override {
+    bool isTuple = false;
```

Comment:
Is it better to write a helper here? This block looks identical between the two ops

Reply:
That is a good point. Done.